app/models/concerns/denormalizable.rb

```diff
@@ -0,0 +1,131 @@
+      case from
+      in Symbol => association_name if reflection = reflect_on_association(association_name)
+        prefixed_attribute_name = case prefix
+                                  when true
+                                    "#{association_name}_#{attribute_name}"
+                                  when Symbol,
+                                       String
+                                    "#{prefix}_#{attribute_name}"
+                                  else
+                                    attribute_name.to_s
+                                  end
+
+        unless reflection.collection?
+          after_initialize  -> { write_denormalized_attribute_from_unpersisted_record(association_name, attribute_name, prefixed_attribute_name) }, if: :"#{reflection.foreign_key}_changed?"
+          before_validation -> { write_denormalized_attribute_from_unpersisted_record(association_name, attribute_name, prefixed_attribute_name) }, if: :"#{reflection.foreign_key}_changed?", on: :create
+          before_create     -> { write_denormalized_attribute_from_persisted_record(association_name, attribute_name, prefixed_attribute_name) },   if: :"#{reflection.foreign_key}_changed?"
+          before_update     -> { write_denormalized_attribute_from_persisted_record(association_name, attribute_name, prefixed_attribute_name) },   if: :"#{reflection.foreign_key}_changed?"
+        else
+          raise ArgumentError, "must be a singular association: #{association_name.inspect}"
+        end
+      else
+        raise ArgumentError, "invalid :from association: #{from.inspect}"
+      end
+    end
+
+    def instrument_denormalized_attribute_to(attribute_name, to:, prefix:)
+      case to
+      in Symbol => association_name if reflection = reflect_on_association(association_name)
+        prefixed_attribute_name = case prefix
+                                  when true
+                                    "#{name}_#{attribute_name}"
+                                  when Symbol,
+                                       String
+                                    "#{prefix}_#{attribute_name}"
+                                  else
+                                    attribute_name.to_s
+                                  end
+
+        # FIXME(ezekg) Set to nil on destroy unless the association is dependent?
+        if reflection.collection?
+          after_initialize  -> { write_denormalized_attribute_to_unpersisted_relation(association_name, prefixed_attribute_name, attribute_name) }, if: :"#{attribute_name}_changed?"
+          before_validation -> { write_denormalized_attribute_to_unpersisted_relation(association_name, prefixed_attribute_name, attribute_name) }, if: :"#{attribute_name}_changed?", on: :create
+          before_create     -> { write_denormalized_attribute_to_persisted_relation(association_name, prefixed_attribute_name, attribute_name) },   if: :"#{attribute_name}_changed?"
```

Comment:
I think the `before_create` should denormalize to unpersisted records, and an `after_create` to persisted records (in case of e.g. denormalizing `id`).